client/virtual_channel_happy_test.go

```diff
@@ -0,0 +1,203 @@
+				}
+			default:
+				t.Fatalf("invalid channel proposal: %v", cp)
+			}
+		},
+	}
+	go ingrid.Client.Handle(proposalHandlerIngrid, proposalHandlerIngrid)
+
+	// Establish ledger channel between Alice and Ingrid.
+	initAllocAlice := channel.Allocation{
+		Assets:   []channel.Asset{asset},
+		Balances: [][]channel.Bal{initBalsAlice},
+	}
+	lcpAlice, err := client.NewLedgerChannelProposal(
+		challengeDuration,
+		alice.Identity.Address(),
+		&initAllocAlice,
+		[]wire.Address{alice.Identity.Address(), ingrid.Identity.Address()},
+	)
+	if err != nil {
+		t.Fatalf("creating ledger channel proposal: %v", err)
+	}
+
+	chAliceIngrid, err := alice.ProposeChannel(ctx, lcpAlice)
+	if err != nil {
+		t.Fatalf("opening channel between Alice and Ingrid: %v", err)
+	}
+
+	// Establish ledger channel between Bob and Ingrid.
+	initAllocBob := channel.Allocation{
+		Assets:   []channel.Asset{asset},
+		Balances: [][]channel.Bal{initBalsBob},
+	}
+	lcpBob, err := client.NewLedgerChannelProposal(
+		challengeDuration,
+		bob.Identity.Address(),
+		&initAllocBob,
+		[]wire.Address{bob.Identity.Address(), ingrid.Identity.Address()},
+	)
+	if err != nil {
+		t.Fatalf("creating ledger channel proposal: %v", err)
+	}
+
+	chBobIngrid, err := bob.ProposeChannel(ctx, lcpBob)
+	if err != nil {
+		t.Fatalf("opening channel between Bob and Ingrid: %v", err)
+	}
+
+	// Setup Bob's proposal handler.
+	chAliceBobBob := make(chan *client.Channel, 1)
+	proposalHandlerBob := &FunctionProposalHandler{
+		openingProposalHandler: func(cp client.ChannelProposal, pr *client.ProposalResponder) {
+			switch cp := cp.(type) {
+			case *client.VirtualChannelProposal:
+				_chAliceBobBob, err := pr.Accept(ctx, cp.Accept(bob.Identity.Address()))
+				if err != nil {
+					t.Fatalf("accepting virtual channel proposal: %v", err)
+				}
+				chAliceBobBob <- _chAliceBobBob
+			default:
+				t.Fatalf("invalid channel proposal: %v", cp)
+			}
+		},
+		updateProposalHandler: func(s *channel.State, cu client.ChannelUpdate, ur *client.UpdateResponder) {
+			ur.Accept(ctx)
+		},
+	}
+	go bob.Client.Handle(proposalHandlerBob, proposalHandlerBob)
+
+	// Establish virtual channel between Alice and Bob via Ingrid.
+	initAllocVirtual := channel.Allocation{
+		Assets:   []channel.Asset{asset},
+		Balances: [][]channel.Bal{initBalsVirtual},
+	}
+	indexMapAlice := []channel.Index{0, 1}
+	indexMapBob := []channel.Index{1, 0}
+	vcp, err := client.NewVirtualChannelProposal(
+		challengeDuration,
+		alice.Identity.Address(),
+		&initAllocVirtual,
+		[]wire.Address{alice.Identity.Address(), bob.Identity.Address()},
+		[]channel.ID{chAliceIngrid.ID(), chBobIngrid.ID()},
+		[][]channel.Index{indexMapAlice, indexMapBob},
+	)
+	if err != nil {
+		t.Fatalf("creating virtual channel proposal: %v", err)
+	}
+
+	chAliceBobAlice, err := alice.ProposeChannel(ctx, vcp)
+	if err != nil {
+		t.Fatalf("opening channel between Alice and Bob: %v", err)
+	}
+
+	err = chAliceBobAlice.UpdateBy(ctx, func(s *channel.State) error {
```

Comment:
Please do _a couple_ of channel updates (at least one) before settling so that the final state is >1. It would be good to test that the hub's channel machines handle skipped version numbers correctly, when the final state is set. I also hinted at this in yesterday's review.